core/src/main/java/org/apache/iceberg/variants/ShreddedArray.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.variants;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.List;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+public class ShreddedArray implements VariantArray {
```

Comment:
Like I commented in the tests, I don't think that "shredded" has much meaning or value for arrays because there's no distinction between values that are shredded or unshredded. The equivalent is a `SerializedArray` (unshredded) and this class.

I would suggest calling this a `ValueArray`. It's a bit hard to name this because it can't be the interface name, `VariantArray` or `Array` since that's too generic. We could use `MutableArray`, but I wouldn't necessarily draw attention to mutability. `ValueArray` seems generic and not great, but I like it better than the other options I've come up with so far.